appcelerator/titanium_mobile
# feat(android)(9_3_X): add camera/microphone foreground service constants

**JIRA:**
https://jira.appcelerator.org/browse/TIMOB-28051

**Summary:**
Adding new Android 11 foreground service constants which allows recording audio/video while the app is in the background.

**Test:**
1. Create a Classic Titanium app project.
2. Copy below "tiapp.xml" settings.
3. Copy below "AudioRecorderServices.js" file to project's "Resources" folder.
4. Use below "app.js" file.
5. Build and run on Android 11.
6. If app ask for microphone permission, then tap yes.
7. Tap on the Home button.
8. Drag down the top status bar.
9. Wait for notification "Audio Recording Service" to show a microphone icon.
10. Say something to the microphone.
11. Tap on the notification to resume the app and verify it repeats what you've said.

tiapp.xml
```xml
<ti:app>
	<android>
		<manifest>
			<uses-sdk android:targetSdkVersion="30"/>
		</manifest>
		<services>
			<service url="AudioRecorderService.js" android:foregroundServiceType="microphone"/>
		</services>
	</android>
</ti:app>
```

./Resources/AudioRecorderService.js
```javascript
Ti.API.info("@@@ AudioRecorderService started.");
var notificationId = 123;
var notification = Ti.Android.createNotification({
	contentTitle: "Audio Recording Service",
	contentText: "I'm listening...",
	contentIntent: Ti.Android.createPendingIntent({
		intent: Ti.App.Android.launchIntent,
	}),
});
Ti.Android.currentService.foregroundNotify(
	notificationId, notification, Ti.Android.FOREGROUND_SERVICE_TYPE_MICROPHONE);
```

./Resources/app.js
```javascript
var audioRecorder = null;
var audioFile = null

// Start recording from the microphone.
function startRecording() {
	stopRecording();
	audioRecorder = Ti.Media.createAudioRecorder();
	audioRecorder.start();
	var lastReference = audioRecorder;
	setTimeout(function() {
		if (audioRecorder == lastReference) {
			stopRecording();
		}
	}, 5000);
}

// Stop recording from the microphone.
function stopRecording() {
	if (!audioRecorder) {
		return;
	}
	if (audioFile) {
		audioFile.deleteFile();
		audioFile = null;
	}
	audioFile = audioRecorder.stop();
	audioRecorder = null;
}

// Start foreground service which enables background recording.
Ti.Android.startService(Ti.Android.createServiceIntent({
	url: "AudioRecorderService.js",
}));

// Start recording when app is in the background after 5 seconds.
Ti.App.addEventListener("paused", function() {
	setTimeout(function() {
		startRecording();
	}, 5000)
});

// When app is brought back to the foreground, play what was recorded in the background.
Ti.App.addEventListener("resumed", function() {
	stopRecording();
	if (!audioFile) {
		return;
	}
	var audioPlayer = Ti.Media.createAudioPlayer({
		url: audioFile.nativePath,
		allowBackground: true,
	});
	audioPlayer.addEventListener("complete", function() {
		setTimeout(function() {
			audioPlayer.release();
		}, 1);
	});
	audioPlayer.start();
});

// Show window which asks for microphone permissions and displays instructions.
var window = Ti.UI.createWindow();
window.add(Ti.UI.createLabel({
	text: "Tap HOME and wait 5 seconds to start recording.\n\nThen resume this app to hear the recording.",
}));
window.addEventListener("open", function() {
	if (!Ti.Media.hasAudioRecorderPermissions()) {
		Ti.Media.requestAudioRecorderPermissions();
	}
});
window.addEventListener("close", function() {
	stopRecording();
});
window.open();
```
